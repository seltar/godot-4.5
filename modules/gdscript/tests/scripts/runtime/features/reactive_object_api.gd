extends RefCounted

var test_object: Object
var signal_count: int = 0

func test():
	test_object = RefCounted.new()
	
	# Test Object methods for reactive properties
	print("Testing Object reactive property methods")
	
	# Test setting a property as reactive
	test_object.set_property_reactive("test_prop", true)
	print("Is test_prop reactive: ", test_object.is_property_reactive("test_prop"))
	
	# Get the signal name for the reactive property
	var signal_name = test_object.get_property_signal_name("test_prop")
	print("Signal name for test_prop: ", signal_name)
	
	# Connect to the reactive signal
	if signal_name != StringName():
		@warning_ignore("return_value_discarded")
		test_object.connect(str(signal_name), _on_property_changed)
	
	# Set the property via metadata (should trigger reactive signal)
	test_object.set_meta("test_prop", "initial_value")
	test_object.set_meta("test_prop", "changed_value")
	
	# Test removing reactive property
	test_object.set_property_reactive("test_prop", false)
	print("Is test_prop reactive after removal: ", test_object.is_property_reactive("test_prop"))
	
	# This should not trigger signal anymore
	test_object.set_meta("test_prop", "final_value")
	
	print("Total signals received: ", signal_count)

func _on_property_changed(old_value, new_value):
	signal_count += 1
	print("Property changed signal: ", old_value, " -> ", new_value)
