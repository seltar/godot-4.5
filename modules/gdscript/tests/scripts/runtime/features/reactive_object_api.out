GDTEST_OK
Testing Object reactive property methods
Is test_prop reactive: true
Signal name for test_prop: test_prop_changed
Property changed signal: <null> -> initial_value
Property changed signal: initial_value -> changed_value
Is test_prop reactive after removal: false
Total signals received: 2
